contracts/PSM/BasePSM.sol

```diff
@@ -0,0 +1,144 @@
+/// Handles basic reserve token transfers as well as minting R.
+/// PSM is also ERC20 that should be added to PositionManager as a collateral with 100% min CR.
+/// PSM ERC20 should have constant price feed returning value of 1e18.
+/// Min CR of 100% and price of 1e18 menas there cannot be liquidations of such collateral.
+/// On top of this, `SplitLiquidationCollateral.split` reverts to make sure no liquidations are possible.
+abstract contract BasePSM is IPSM, ERC20, Ownable2Step {
+    using SafeERC20 for IERC20;
+    using Fixed256x18 for uint256;
+
+    IERC20 public immutable override reserveToken;
+    IRToken public immutable override rToken;
+    IPositionManager public immutable override positionManager;
+    IPSMFeeCalculator public override feeCalculator;
+
+    constructor(
+        IERC20 reserveToken_,
+        IRToken rToken_,
+        string memory name_,
+        string memory symbol_,
+        IPSMFeeCalculator feeCalculator_
+    )
+        ERC20(name_, symbol_)
+    {
+        if (address(reserveToken_) == address(0)) {
+            revert ZeroInputProvided();
+        }
+
+        reserveToken = reserveToken_;
+        rToken = rToken_;
+        positionManager = IPositionManager(rToken_.positionManager());
+
+        setFeeCalculator(feeCalculator_);
+
+        _approve(address(this), address(positionManager), type(uint256).max);
+    }
+
+    function buyR(uint256 reserveAmount, uint256 minReturn) external returns (uint256 rAmountOut) {
+        if (reserveAmount == 0 || minReturn == 0) {
+            revert ZeroInputProvided();
+        }
+
+        reserveToken.safeTransferFrom(msg.sender, address(this), reserveAmount);
+        _depositReserveToken(reserveAmount);
+
+        rAmountOut = reserveAmount - feeCalculator.calculateFee(reserveAmount, true);
+        if (rAmountOut < minReturn) {
+            revert ReturnLessThanMinimum(rAmountOut, minReturn);
+        }
+        _mintR(msg.sender, rAmountOut);
+    }
+
+    function buyReserveToken(uint256 rAmount, uint256 minReturn) external returns (uint256 reserveAmountOut) {
+        if (rAmount == 0 || minReturn == 0) {
+            revert ZeroInputProvided();
+        }
+
+        _burnR(msg.sender, rAmount);
+
+        reserveAmountOut = rAmount - feeCalculator.calculateFee(rAmount, false);
+        if (reserveAmountOut < minReturn) {
+            revert ReturnLessThanMinimum(reserveAmountOut, minReturn);
+        }
+        _withdrawReserveToken(reserveAmountOut);
+        reserveToken.safeTransfer(msg.sender, reserveAmountOut);
+    }
+
+    function withdrawReserve(address to, uint256 amount) external override onlyOwner {
+        if (to == address(0) && amount == 0) {
+            revert ZeroInputProvided();
+        }
+        _withdrawReserveToken(amount);
+        reserveToken.safeTransfer(to, amount);
+    }
+
+    function recoverTokens(IERC20 token, address to, uint256 amount) external override onlyOwner {
```

Comment:
check `to != address(0)` (or just transfer to msg.sender

Reply:
`msg.sender` could be `onlyOwner`, so, always would be transferred to the owner. So, maybe only `to != address(0)`, but because it is `onlyOwner` maybe and that is not needed.

Reply:
<img src="https://giphy.com/gifs/curbyourenthusiasm-hbo-curb-your-enthusiasm-gfxZmGHj5EhE2ts2SP" width="40" height="40" />